src/formats/MOL2.cpp

```diff
@@ -0,0 +1,298 @@
+        unsigned long id, resid;
+        char atom_name[32], atom_type[32], res_name[32];
+        double x, y, z;
+        double charge = 0;
+
+        if (charges) {
+            scan(line, " %lu %31s %lf %lf %lf %31s %lu %31s %lf",
+                &id, &atom_name[0], &x, &y, &z, &atom_type[0], &resid, &res_name[0], &charge
+            );
+        } else {
+            scan(line, " %lu %31s %lf %lf %lf %31s %lu %31s",
+                &id, &atom_name[0], &x, &y, &z, &atom_type[0], &resid, &res_name[0]
+            );
+        }
+
+        Atom atom(atom_name);
+        frame.add_atom(std::move(atom), Vector3D(x, y, z));
+        size_t current_atom = frame.size() - 1;
+        if (charges) {
+            frame[current_atom].set_charge(charge);
+        }
+        frame[current_atom].set_type(atom_type);        
+
+        if (residues_.find(resid) == residues_.end()) {
+            Residue residue(res_name, resid);
+            residue.add_atom(current_atom);
+
+            residues_.insert({resid, residue});
+        } else {
+            // Just add this atom to the residue
+            residues_.at(resid).add_atom(current_atom);
+        }
+
+    }
+}
+
+void MOL2Format::read_bonds(Frame& frame, size_t nbonds) {
+    auto lines = file_->readlines(nbonds);
+
+    for (const auto& line : lines) {
+        unsigned long id, id_1, id_2;
+
+        scan(line, " %lu %lu %lu", &id, &id_1, &id_2);
+
+        // MOL2 is 1 index-based, not 0
+        --id_1;
+        --id_2;
+
+        if (id_1 >= frame.size() || id_2 >= frame.size()) {
+            throw format_error("Connect ID '{}' or '{}' larger than number of atoms '{}'",
+                id_1, id_2, frame.size()
+            );
+        }
+
+        frame.add_bond(id_1, id_2);
+    }
+}
+
+std::streampos read_until(TextFile& file, const std::string& tag) {
+    while (!file.eof()) {
+        std::streampos pos = file.tellg();
+        if (file.readline().substr(0, tag.length()) == tag) {
+            return pos;
+        }
+    }
+
+    throw file_error("File ended before tag '{}' was found", tag);
+}
+
+std::streampos forward(TextFile& file) {
+    if (!file) {return std::streampos(-1);}
+    while (!file.eof()) {
+        try {
+            std::streampos pos = read_until(file, "@<TRIPOS>MOLECULE");
+            file.readline(); // Skip a line
+            auto line = file.readline();
+
+            const auto counts = split(line, ' ');
+
+            long long natoms, nbonds = 0;
+            natoms = string2longlong(counts[0]);
+
+            if (natoms < 0) {
+                throw format_error(
+                    "number of atoms can not be negative in '{}'", file.filename()
+                );
+            }
+
+            if (counts.size() >= 2) {
+                nbonds = string2longlong(counts[1]);
+            }
+
+            read_until(file, "@<TRIPOS>ATOM");
+            file.readlines(static_cast<size_t>(natoms));
+
+            if (nbonds > 0) {
```

Comment:
what happens if nbonds is 0 or negative?

Reply:
At least a warning would be good, as it might help to understand why there is an error.